### ISSUE TITLE
Please use &self on context.make_current(), otherwise managing multiple contexts is painful

### ISSUE THREAD

**Issue by fschutt** [rust-windowing/glutin#1183]
A problem in `glutin-0.22.0-alpha` is that the `context.make_current()` *consumes* the entire context when making it current. This is a problem the second you want to store the context somewhere.

Please make `.make_current()` take a `&self`. I know that you probably wanted to have a type-safe way so that `.swap_buffers()` can only be called on a current context - while this is a nice idea, it fails in practice, swapping buffers on non-current windows hasn't been such a huge problem.

```rust
struct Window {
    context: WindowedContext<NotCurrent>,
}

fn render(window: Window) {
    loop {
         // calling .make_current() moves out of "window.context" here, so this doesn't compile
         let current_context = window.context.make_current().unwrap();
    }
}
```

**Comment by ZeGentzy** [comment]
This new api has been the case since v0.21.0, not v0.22.0-alpha1, no complaints related to this new api were filed from RC1 to RC3.

Have you taken a look at [the migration guide](https://gentz.rocks/posts/glutin-v0-21-0-migration-guide/)?

**Comment by fschutt** [comment]
Yeah, I'm upgrading from 0.20. The migration guide doesn't cover handling multiple contexts. Basically my problem is this: I have a "root context" that all windows share (using shared lists). But each window still has it's own `WindowedContext`. The renderer is attached to the root context, it draws onto a texture, and the texture is then (since OpenGL doesn't allow sharing framebuffers)

The migration docs say that I should just do this:

```rust
struct Window<T: ContextCurrentState> {
    context: WindowedContext<T>,
}

impl Window<NotCurrent> {
    fn make_current(self) -> Self {
        Self { context: self.context.make_current().unwrap() }
    }
}
```

That's what the migration guide says the solution to my problem is. So (a) now I have to carry the `T: ContextCurrentState` everywhere (the generics bubble up into my library types), (b) every time I want to make the context current I have to *construct an entirely new window*.

Why is this solved using generics? Why isn't this just the simpler solution:

```rust
struct CurrentContext {
    handle: *mut c_void,
}

struct NotCurrentContext {
    handle: *mut c_void,
}

enum WindowContext {
    Current(CurrentContext),
    NotCurrent(NotCurrentContext),
}

impl WindowContext {
    pub fn make_current(&mut self) {
        match self {
            Current(c) => { },
            NotCurrent(NotCurrentContext { raw_handle })) => { 
              // self.os_window.make_current(); // - the call to the OS to make the context current
              *self = Current(CurrentContext { raw_handle });
        }
    }

    pub fn get_current_context(&self) -> Option<&CurrentContext> {
        match self {
            Current(c) => Some(c),
             _ => None,
        }
    }
}

impl CurrentContext {
    pub fn swap_buffers(&self) {
       // self.raw_handle.swap_buffers() // or whatever the OS call is
    }
}
```

... or something like this. No generics needed, still type-safe, since the user needs to check first whether the context is actually current. The current API is awful. I just want to make the GL context current, so that I can draw a texture to the backbuffer of a window. Previously it used to be just one line (`context.make_current()`), now it's [several hundred lines](https://github.com/rust-windowing/glutin/blob/dcdfc01bbbf407cc08930111995181066803ccd7/glutin_examples/examples/support/mod.rs#L169) because of the ownership issues.

Here's what I'm trying to port: https://github.com/maps4print/azul/blob/6763ca72f711e5874bf9d259a0eea9fe5f32f16c/azul/app.rs#L1331-L1335. This is impossible to port with the current API, since the "root window" gets moved out, when I make the context current, and I can't constantly destroy-and-recreate contexts just to work around a mediocre API.

**Comment by ZeGentzy** [comment]
The context tracker code isn't the finest code I've ever written. Please don't take it has the beacon of high quality glutin code and I hope will be improved for v0.23.0

Nevertheless, you can use the current api like this:
```rust
enum ContextInner {
    Not(WindowedContext<NotCurrent>),
    Possibly(WindowedContext<PossiblyCurrent>),
}

struct Context {
    inner: Takeable<ContextInner>,
}

impl Context {
    pub unsafe fn make_current(&mut self) {
        let inner = Takeable::take(&mut self.inner);
        let context = match inner {
            ContextInner::Possibly(c) => c.make_current().unwrap(),
            ContextInner::Not(c) => c.make_current().unwrap(),
        };
        Takeable::insert(&mut self.inner, ContextInner::Possibly(context));
    }

    pub fn get_current_context(&self) -> Option<&WindowedContext<PossiblyCurrent>> {
        match *self.inner {
            ContextInner::Possibly(ref c) => Some(c),
            _ => None,
        }
    }
}
```

**Comment by ZeGentzy** [comment]
> (b) every time I want to make the context current I have to construct an entirely new window.

You're not *really* constructing a new window. Your constructing a new struct, yes, but it's the same window.

>  now it's several hundred lines because of the ownership issues.
It's not like that cause ownership reasons, it's like that cause it's shitty.

**Comment by fschutt** [comment]
Okay, I think I solved it in the way you posted.

```rust
enum ContextState {
    Current(WindowedContext<PossiblyCurrent>),
    NotCurrent(WindowedContext<NotCurrent>),
}

impl ContextState {
    pub fn make_current(&mut self) {

        use std::mem;
        use self::ContextState::*;

        let mut self_mem = unsafe { mem::zeroed::<Self>() };
        mem::swap(&mut self_mem, self); // self -> self_mem

        let mut new_state = match self_mem {
            Current(c) => Current(c),
            NotCurrent(nc) => Current(unsafe { nc.make_current().unwrap() }), // TODO: swap back on Err()
        };

        mem::swap(&mut new_state, self); // self_mem -> self
        mem::forget(new_state); // can't call the destructor on zeroed memory
    }
}
```

**Comment by ZeGentzy** [comment]
But why? That code will cause UB. If `make_current` panics, the `mem:zeroed::<Self>()` will be dropped!
Also, there is a `Possibly` in `PossiblyCurrent` for a reason. That code should also be calling `make_current` on the `WindowedContext<PossiblyCurrent>`.

**Comment by fschutt** [comment]
>  If make_current panics, the mem:zeroed::<Self>() will be dropped!

I know, that's why I put the `TODO` comment in there. It works for now, I'll fix it later.

**Comment by ZeGentzy** [comment]
But why use unsafe stuff, when you can use `Option` or `Takeable` from takeable-option?

**Comment by fschutt** [comment]
Because of the cost of extra dependencies. It's already hard enough convincing upstream maintainers not to add random crates that import 100 sub-dependencies. Build times are already bad enough, it's not worth it for a 3-line solution.

That, and I generally dislike `Takeable<Iterable<Filterable<X>>>` monstrosities. Keep it simple.

**Comment by ZeGentzy** [comment]
`Option` is in the std libraries (a.k.a. no deps), while takeable-option only depends on the std libraries.

**Comment by dvc94ch** [comment]
The problem with the option solution is that it requires a mutex when the context is shared accross threads. Can we please reopen this issue?

**Comment by goddessfreya** [comment]
@dvc94ch For whatever it's worth, v0.23 does change it back to a `&self`, see https://github.com/rust-windowing/glutin/pull/1239/files#diff-0b8348625b3d5cb59bd3ca9507a92906R24

**Comment by goddessfreya** [comment]
Implemented for v0.24. See #1271. Closing.
